hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+    MINE = 4
+    SAVEDLABEL = 5
+    LABEL_CHOICES = (
+        (LABEL, 'Label'),
+        (FOLDER, 'Folder'),
+        (FAVORITE, 'Favorite'),
+        (MINE, 'Mine'),
+        (SAVEDLABEL, 'Saved Label')
+    )
+
+####################################
+# user access to resources
+####################################
+
+
+class UserResourceLabels(models.Model):
+    """ Labels of a user for a resource
+
+    This model stores labels of an individual user, like an access
+    control list; There are several kinds of labels documented in LabelCodes.
+    These are similar in implementation but differ in semantics.
+    """
+
+    kind = models.IntegerField(choices=LabelCodes.LABEL_CHOICES,
+                               editable=False,
+                               default=LabelCodes.LABEL)
+    # provenance of label start
+    start = models.DateTimeField(editable=False, auto_now=True)
+
+    # if end is non-null, then the label is no longer valid. Not absolutely sure we need this.
+    # end  = models.DateTimeField(editable=False, null=True, auto_now=False)
+
+    # Utilize an underlying class of User to hold labels for resources.
+    # This creates an implicit query User.ulabels that represents all labels
+    # for a user.
+    ulabels = models.ForeignKey('UserLabels', null=False, editable=False,
+                                related_name='ul2url',         # unused but must be defined and unique
+                                help_text='user assigning a label')
+    # There is a basic design question as to whether the target of a label should be a
+    # Resource proper, or an annotation class that contains mainly methods for the resource.
+    # Should this be 'BaseResource' or 'ResourceLabels'?
+    rlabels = models.ForeignKey('ResourceLabels', null=True, editable=False,
+                                related_name="rl2url",     # unused but must be defined and unique
+                                help_text='resource to which a label applies')
+    label = models.TextField()
+
+
+class UserLabels(models.Model):
+    """
+    Projection class puts methods and content inside basic User object
+    so that one can access things easily from that context.
+    This model is injected into the BaseResource as the related name "ulabels".
+    Thus for an User u, u.ulabels is this model.
+    """
+
+    user = models.OneToOneField(User,
+                                editable=False,
+                                null=True,
+                                related_name='ulabels',  # induced field in User class.
+                                related_query_name='ulabels')
+
+    ##########################################
+    ##########################################
+    # PUBLIC FUNCTIONS: resources
+    ##########################################
+    ##########################################
+
+    ##########################################
+    # get various aspects of labeling
+    ##########################################
```

Comment:
Comment blocks again unnecessary. Docstrings should be the canonical documentation of what these methods are and how they relate.